Update: Use a privacy-compliant pattern in Next.js cookie banner tutorial (#7464)
PostHog/posthog.com: contents/tutorials/nextjs-cookie-banner.md

}
```

After setting this up and running `npm run dev`, PostHog starts autocapturing events, but a PostHog-related cookie is set for the user without their consent.

![Cookie set](../images/tutorials/nextjs-cookie-banner/cookie.png)

## Creating a cookie banner

Create another file in the `app` folder named `banner.js` for the banner component with a bit of text explaining cookies and buttons to accept or decline.

> Importantly, to avoid a hydration error, we must check if the frontend has mounted and only show the component if so. We can use `useState` and `useEffect` to do this. Together, this looks like this:

Reviewer: we can remove this if the initial state is true and not false?

Author: If we set the initial state to true, it flickers the cookie banner when hidden, showing up briefly before going away.

'use client';
import { useEffect, useState } from "react";

export function Banner() {

>   const [showBanner, setShowBanner] = useState(false);

Reviewer: just set the initial state to true for now? Then we can remove the bit about the hydration error

Author: As far as I remember, the hydration error happens if you don't use `useState` at all.

> 

Reviewer: I cant comment on the line above for some reason, but change the sentence to:

`This creates an ugly but functional cookie banner at the bottom of our site. You can customize and style it how you like.`

    const storedConsent = localStorage.getItem(consentKey);
    if (storedConsent) {
      setConsent(storedConsent);
    }
  }, []);

  const updateConsent = (newConsent) => {
    // Only update consent if the user hasn't made a decision yet
    if (['granted', 'denied'].includes(consent)) {
      return;
    }
    setConsent(newConsent);
    if (typeof window !== 'undefined') {
      localStorage.setItem(consentKey, newConsent);
    }
  };

  return (
    <ConsentContext.Provider value={{ consent, updateConsent, setConsent }}>
      {children}
    </ConsentContext.Provider>
  );
};
```

Finally, we add the consent provider to our `layout.js` file and wrap our app in it.

```js
import './globals.css'
import { PHProvider } from './providers'
import { ConsentProvider } from './consent'
import Banner from './banner'

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <ConsentProvider>
        <PHProvider>
          <body>
            {children}
            <Banner />
          </body>
        </PHProvider>
      </ConsentProvider>
    </html>
  )
}
```

## Setting the consent state with the cookie banner

> We want the cookie banner to control the consent state so we can use it to initialize PostHog either with or without cookies. We also want to use it to hide the banner when the user makes a decision.

Reviewer: ```suggestion
Depending on the consent state, we want to initialize PostHog either with or without cookies. We also want to use it to hide the banner when the user makes a decision.
```
